src/battle.rs

```diff
@@ -505,6 +522,52 @@ pub (crate) fn attack_button_handler (
     }
 }
 
+fn calculate_damage(player: &MonsterBundle, player_action: usize, 
+    enemy: &MonsterBundle, enemy_action: usize) -> (usize, usize) {
+    if (player_action == 1 || enemy_action == 1) {
+        // if either side defends this turn will not have any damage on either side
+        return (0, 0);
+    }
+    // More actions can be added later, we can also consider decoupling the actions from the damage
+    let mut result = (0,0);
+    // player attacks
+    // If our attack is less than the enemy's defense, we do 0 damage
+    if player.stg.atk <= enemy.def.def {
+        result.0 = 0;
+    } else {
+        // if we have damage, we do that much damage
+        // I've only implemented crits for now, dodge and element can follow
+        result.0 = player.stg.atk - enemy.def.def;
+        if player.stg.crt <= enemy.def.crt_res {
+            result.0 = result.0;
+        } else {
+            // calculate crit chance and apply crit damage
+            let mut crit_chance = player.stg.crt - enemy.def.crt_res;
+            let crit = rand::thread_rng().gen_range(0..=100);
+            if crit <= crit_chance {
+                result.0 *= player.stg.crt_dmg;
+            }
+        }
+    }
+    // same for enemy
```

Comment:
Could these be refactored into macros to clean up and practice DRY?

Reply:
Feel free to try. I did not look into these.